contracts/tokenomics/maker/src/contract.rs

```diff
@@ -452,12 +485,43 @@ fn distribute_astro(deps: DepsMut, env: Env, info: MessageInfo) -> Result<Respon
 /// # Params
 /// * **deps** is the object of type [`DepsMut`].
 ///
+/// * **env** is the object of type [`Env`].
+///
 /// * **cfg** is the object of type [`Config`].
 ///
 /// * **amount** is the object of type [`Uint128`].
-fn distribute(deps: Deps, cfg: &Config, amount: Uint128) -> Result<Vec<SubMsg>, ContractError> {
+fn distribute(
+    deps: DepsMut,
+    env: Env,
+    cfg: &mut Config,
+    mut amount: Uint128,
+) -> Result<Vec<SubMsg>, ContractError> {
     let mut result = vec![];
 
+    if !cfg.rewards_enabled {
+        cfg.pre_upgrade_astro_amount = amount;
+        CONFIG.save(deps.storage, cfg)?;
+        return Ok(result);
+    } else if !cfg.pre_upgrade_astro_amount.is_zero() {
+        let remainder_reward = cfg.remainder_reward.unwrap_or(cfg.pre_upgrade_astro_amount);
+        let blocks_passed = Uint128::from(env.block.height - cfg.last_distribution_block);
+        let astro_distribution_portion =
+            cfg.pre_upgrade_astro_amount / Uint128::from(cfg.pre_upgrade_blocks);
+        let current_distribution =
+            max(blocks_passed * astro_distribution_portion, remainder_reward);
+
+        // subtract pre-upgrade astro amount
+        amount -= cfg.pre_upgrade_astro_amount;
+
+        // add a portion of reward
+        amount += current_distribution;
+
+        // reduce the number of pre-upgrade astro amount
+        cfg.remainder_reward = Some(remainder_reward - current_distribution);
```

Comment:
we should safety subtraction, cuz current_distribution can be greater than remainder_reward

Reply:
Fixed. Thanks